### PR TITLE
FIX: Select appropriate action/action map when right clicking (ISX-1743).

### PR DESCRIPTION
**Pull request by graham-huws**
### Description

Fix for [ISX-1743](https://jira.unity3d.com/browse/ISX-1743) - Right clicks would not change the selected item, giving a misleading context menu.

### Changes made

Fixed: Selection not changing when right clicking an Action Map or Action

### Notes

Unfortunately, right clicking a different Action and selecting Rename does nothing until right clicking a second time (on Mac) - There seems to be some voodoo about this which probably needs a #devs-uitoolkit wizard. The renameTextfield.SelectAll() call doesn't seem to do anything, and the Focus() call afterwards fails too. I wondered if this may be related to [this (closed as won't fix) bug](https://jira.unity3d.com/browse/UUM-1571) from a while ago, though changing the event to MouseDown didn't help in my case.

As is, I think it makes sense to push this PR ahead as-is, and a new lower prio bug should be logged for this remaining issue.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Review by ekcoh — COMMENTED** [review]
In general LGTM, however it sounds like the null behavior hasn't really been understood completely. When @ritamerkl is available again it would be great to get your perspective on this. Will likely provide additional feedback once I have been able to test the branch.

**Review by jamesmcgill — APPROVED** [review]
Looks like good changes to me (simpler/less code).
I agree we should get this merged rather than holding out for all issues to be resolved.
However I'll let @ritamerkl do the merge after she gets a chance to review this.

**Comment by Pauliusd01** [comment]
Checking

**Comment by Pauliusd01** [comment]
Besides the mentioned 2nd click being needed issue, this also introduces a new issue with multiple Input Action Windows present - if you select something in one window and then right-click in the other, the context menu will always do things on the first window's selection, no matter how many times you left click or right click in the second. (Adding a video of me left/right clicking between windows in your PR and then in current Develop) I'm not sure if this is as an easy issue to ignore as the one you already found

https://github.com/Unity-Technologies/InputSystem/assets/54306142/84983acd-bda3-46fa-9b4c-5a8d03157871

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
My recommendation is leaning more to fix the multi window issue before merging but if someone else says It's fine I can change the status

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/InputActionsTreeViewItem.cs
Comment (ritamerkl) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/InputActionsTreeViewItem.cs]: this is ok in my point of view as long as it is still working the intended way - even though I don't understand the context to the problem of the PR

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ritamerkl) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:450]: I am a bit worried about setting selections like this. Is it even the expected behaviour to select things on right click? If then we should change it for actionmaps, actions and bindings to keep it consistent. Also I have the feeling there is a simpler way to do that then this, in particular looking at the while loop. (maybe there is even a way to allow right click selection in ActionsTreeView?) 
Also did we contact the UIToolkit team members here? Maybe they have a recommandation for this case.

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: I would suggest adding some context here, i.e.

"Fixed selection not changing when right-clicking an Action Map or Action in the Project Settings Input Action Editor."

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:450]: I am also suprised there is not a out-of-the-box solution for this in UITK that would provide consistent behavior across the editor. I would rather keep this behavior instead of introducing inconsistency with the rest of the editor. Could we reach out to UITK team as suggested by Rita and possible find similar examples in the editor?

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:450]: Some examples I found by quickly testing in the editor:
- Tags and layers in Project Settings list boxes (might be separate controls though), change focus/selection on right click
- Noticed that list in Package Manager settings do not select on right click but neither has a context menu either.
- Tree view in Project Settings (left panel) select on right click but do not have a context menu
- Project Hierarchy (probably most similar example), select on right click and has a context menu, so maybe UITK team can provide guidance on how that was achieved? Custom or some common mechanism? Also Game Object Hierarchy behaves this way

Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:450]: Thanks for comments both!

> Is it even the expected behaviour to select things on right click?

I would think so - as well as the project/gameobject areas mentioned by Håkan, it's standard behaviour across operating systems, so I'd expect it to work here too.

> Also I have the feeling there is a simpler way to do that then this, in particular looking at the while loop.

Same! 😆 Maybe a question for #devs-uitoolkit if nobody has a suggestion? I'm still very new to it all.

Regarding asking UITK team about rightclick - the method here is one suggested by someone on their team just a few months ago, for the same issue: https://unity.slack.com/archives/C3414V4UV/p1693320293138709?thread_ts=1693306035.515169&cid=C3414V4UV and https://unity.slack.com/archives/C3414V4UV/p1694008889654529

Project and Hierarchy views are both IMGUI unfortunately, so can't make use of their methods 😞

Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:450]: > I would think so - as well as the project/gameobject areas mentioned by Håkan, it's standard behaviour across operating systems, so I'd expect it to work here too.

Okay, wasn't aware of this, but that simplifies things!

I think #devs-uitoolkit is a good place to ask. Also both threads seem to be helpful, especially the trickle down recommendation for catching events before the context menu rightclick triggers.

**Review by Pauliusd01 — APPROVED** [review]
LGTM, the issue I've mentioned is fixed.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ritamerkl) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:443]: how about use panel.Pick(evt.mousePosition)? that could save a while loop here

**Review by ritamerkl — APPROVED** [review]
Looks good to me, I saw the [thread](https://unity.slack.com/archives/C3414V4UV/p1693320293138709?thread_ts=1693306035.515169&cid=C3414V4UV) about the selecting on right click in devs-uitoolkit. I just left one last comment on the while loop. I didn't test it, ok for me if @Pauliusd01 approved it. The [issues](https://github.com/Unity-Technologies/InputSystem/pull/1806#discussion_r1444683114) Håkan mentioned are to fix before though!

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:443]: I tried that - unfortunately it just returns the same element as evt.target 😞
I think the issue is that in the scrollview, the hierarchy _should_ have everything descending from InputActionsTreeViewItem, which would mean I could get ancestor of type...

**Comment by graham-huws** [comment]
@Pauliusd01 Could you have a check with the "add binding" button to make sure it's behaving - I added what should be a small improvement in 80d5916cd52fd5c3acccb6bfb7f0944c504aeebc.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:443]: right, that would be a better option. If this is not working we should at least introduce a safety break in this loop, if no element is found. Something like element.parent !=null

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
https://github.com/Unity-Technologies/InputSystem/assets/54306142/44d56755-7cf7-4435-a069-1422f5e2986f

Add binding on rightclick is broken. Steps are: open 2 input actions windows -> right click one of them and click add binding -> if the binding is added correctly then repeat the previous step in the other window -> now the binding should be added incorrectly

Also, sometimes I get this error thrown but it does not seem to be consistent, might be related to the number of actions in the asset:
```
Cannot add binding without an action and action map selected
UnityEditor.GenericMenu:CatchMenu (object,string[],int) (at D:/Gitrepo/unity/Editor/Mono/GUI/GenericMenu.cs:127)
```

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (graham-huws-unity) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:443]: I _think_ in theory that shouldn't be possible, but sensible idea - added in b9f698fd36f6995569cb232ba3e6d26e0d1820bf.

**Comment by graham-huws** [comment]
> Add binding on rightclick is broken. 

This seems to be broken on develop for me, going between windows - it gives "Cannot add binding without an action and action map selected" every time.

**Review by Pauliusd01 — APPROVED** [review]

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:450]: It think we should file a feature request for this. This might be a workaround now but it will for sure be needed  down the road for Project Hierarchy and then it would be nice to at least have a ticket to track/monitor.

**Review by ekcoh — APPROVED** [review]
If the workaround works then we should get it in. However I would advice we file a feature request on the UI team for this functionality since I would like to avoid having custom logic to handle such a common use-case. if there was a thread for this @graham-huws DM me and we can sort it out there.

For testing I would advice testing on multiple OS, e.g. Win + macOS since these behavior might be inconsistent (based on related bugs)

**Pull request merged**
